book/src/developers/designs/identity-verification.md

```diff
@@ -0,0 +1,160 @@
+
+Implements #337
+
+### Goal:
+
+Providing a way to allow 2 subject to authenticate each other without prior knowledge of each other
+
+### Assumptions:
+
+Both subjects will have to be connected to a Kanidm server (not necessarily the same or under the same domain), and the servers will have to mutually trust each other; the trust bond must be established beforehand.
+
+### High level overview:
+
+The idea is to associate a [ECDH](https://docs.rs/openssl/latest/openssl/pkey_ctx/struct.PkeyCtxRef.html#method.derive_set_peer) public/secret key pair with each user and make each server share a token time and a secret. If two users want to authenticate each other, their private key will be used to encrypt a HMAC hash of the token time and the secret shared between their servers, and the servers will use the public key to decrypt the hash and verify if it matches with the one computed locally. We could also allow users to use their public key to encrypt whatever they want, and then use their private key to decrypt it, providing a more flexible out of band challenge.
+
+### Trust bond between servers:
+
+The way two Kanidm servers create a bond of trust is in many ways similar to the way SAML establishes trust between entities belonging to the same federation.
+Depending on wether the servers belong to the same domain or not, the process will be slightly different.
+Since dealing with different domains would add some degrees of complexity, the initial focus will be on the case where the servers belong to the same domain.
+
+### Servers belonging to the same domain:
+
+```
+┌──────────────────────────────────────────┐
+│Domain A                                  │
+│┌────────────┐              ┌────────────┐│
+││            │              │            ││
+││            │              │            ││
+││            │              │            ││
+││  Server 1  │◀────────────▶│  Server 2  ││
+││            │              │            ││
+││            │              │            ││
+││            │              │            ││
+│└────────────┘              └────────────┘│
+│                                          │
+└──────────────────────────────────────────┘
+```
+
+- Two servers interested in creating such trust bond will perform a ECDH key exchange to create a shared secret, encrypting the communication with TLS
+- a KBKDF, such as the [Scrypt](https://docs.rs/openssl/latest/openssl/pkey_ctx/struct.PkeyCtxRef.html#method.derive_set_peer) KDF will be used to generate a secret key from the shared secret,
+- the secret key will stored in the database of both servers,
+- Finally, they both store the current unix time in the database (which will be referred as _T0_); it will be used for TOTP generation. Every time a new authentication request is received, the _token-time_ will be equal to the number of 30 seconds interval that have passed since _T0_, ie _token-time = (current_unix_time - T0) / 30_.
```

Comment:
We don't have to store the current time at all. We also don't need TLS at all either. The "arrows" here represent that replication *has* occured, but may not be currently occuring. 

Try not to think about this as something where there is any active communication at all.

Server 1 has accounts A and B. A and B both have a private key / public key.

Server 1 then replicates the full DB content to Server 2. Server 2 is now an identical copy.

At this point we *disconnect server 1 and 2* from each other. They can't talk to each other at all!

Person A logs into server 1

Person B logs into server 2

How can they validate each other?

Since server 1 can internally read the private secrets of account A and B it can get the private/public keys and derive the ecdh secrets. 

Server 2 can do the same process.

These secrets are now in memory on server 1 and server 2 and they can be used as the totp seeds.

See how there is no communication outside the original replication event? Because the DB is replicated, each server has everything needed to all arrive at the same results without communication which is absolutely critical in a distributed system like this :)